reconcile/dashdotdb_dvo.py

```diff
@@ -0,0 +1,126 @@
+    def __init__(self, dry_run, thread_pool_size):
+        self.dry_run = dry_run
+        self.thread_pool_size = thread_pool_size
+        self.settings = queries.get_app_interface_settings()
+        secret_reader = SecretReader(settings=self.settings)
+        secret_content = secret_reader.read_all({'path': DASHDOTDB_SECRET})
+        self.dashdotdb_url = secret_content['url']
+        self.dashdotdb_user = secret_content['username']
+        self.dashdotdb_pass = secret_content['password']
+        self.logmarker = "DDDB_DVO:"
+
+    def _post(self, deploymentvalidation):
+        if deploymentvalidation is None:
+            return None
+        cluster = deploymentvalidation['cluster']
+        dvdata = deploymentvalidation['data']
+        LOG.info('%s posting validations for %s', self.logmarker, cluster)
+        response = None
+        if not self.dry_run:
+            endpoint = (f'{self.dashdotdb_url}/api/v1/'
+                        f'deploymentvalidation/{cluster}')
+            response = requests.post(url=endpoint, json=dvdata,
+                                     auth=(self.dashdotdb_user,
+                                           self.dashdotdb_pass))
+            try:
+                response.raise_for_status()
+            except requests.exceptions.HTTPError as details:
+                LOG.error('%s error posting %s - %s',
+                          self.logmarker, cluster, details)
+
+        LOG.info('%s cluster %s synced', self.logmarker, cluster)
+        return response
+
+    def _promget(self, url, query, token=None):
+        uri = 'api/v1/query'
+        url = urljoin('https://'+url, uri)
+        headers = {
+            "accept": "application/json",
+            "Authorization": "Bearer " + token,
+        }
+        params = {'query': query}
+        LOG.info('%s Fetching prom payload from %s?%s',
+                 self.logmarker, url, query)
+        response = requests.get(url,
+                                params=params,
+                                headers=headers,
+                                timeout=(5, 30))
+        try:
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as details:
+            LOG.error('%s error accessing prometheus - %s',
+                      self.logmarker, details)
+
+        response = response.json()
+        # TODO ensure len response == 1
+        # return response['data']['result']
+        return response
+
+    def _get_automationtoken(self, cluster):
+        autotoken_reader = SecretReader(settings=self.settings)
+        autotoken = {'path': "app-sre/creds/kube-configs/" + cluster,
+                     'field': "token"}
+        token = autotoken_reader.read(autotoken)
+        return token
+
+    def _get_deploymentvalidation(self, cluster, validation, oc_map):
+        LOG.debug('%s processing %s, %s', self.logmarker, cluster, validation)
+        cluster_promurl = "prometheus." + cluster + ".devshift.net"
```

Comment:
you'd think so.. but several of the clusters don't have an application prometheusUrl defined (as they are network-isolated). note: the way prometheus urls are managed in grafana is %100 manually editing the secrets file.. and i wanted to aim for the least hands-on approach for maintaining configuration. this seemed like an okay way to derive the prometheusUrl per cluster.. with the caveat that some urls are invalid and therefore would not resolve or not connect, and can then be skipped.

i could add the prometheusUrl property to the generic cluster-fetch query, and then skip any clusters where this is none. maybe that's better, though i'm not sure if that has impact to anything consuming that query. note: i wish there was a way to schema+version these queries, to prevent backwards compatibility issues:  
https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/queries.py#L268-L270

wdyt? /cc @apahim

Reply:
I think it is safer than building it manually as app-interface is the source of truth. If it doesn't exist, it should mean that there are no prometheus there...

Reply:
adding property to query.py.. need to retool _get_deploymentvalidation slightly to read that property and skip the cluster if promurl isn't set.

Reply:
rebased this entirely to use data from app-interface. if prometheusUrl isn't set, we skip..  however there's still some cases where the prometheusUrl is set, but can't be reached.. so.. neat.